scheduler/plugin/pool.go

```diff
@@ -0,0 +1,142 @@
+/*
+   Copyright 2023 The Kubernetes Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package wasm
+
+import (
+	"context"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// guestPool manages guest to pod assignments in a scheduling or binding cycle.
+type guestPool[guest any] struct {
+	// newGuest is a function to create a new guest.
+	newGuest func(context.Context) (guest, error)
```

Comment:
By extracting this function and making it a generic type, I was able to find a subtle bug in the test.